extra_data/caldata.py

```diff
@@ -0,0 +1,1016 @@
+            snapshot_at (datetime, date, str or None): Time of database
+                state to look at.
+
+        Returns:
+            None
+        """
+
+        da_to_modno = {data['karabo_da']: modno for modno, data in
+                       self.pdu_by_module_number.items()}
+
+        if not calibrations:
+            # Make sure there are at least empty dictionaries for each
+            # module.
+            for modno in da_to_modno.values():
+                metadata[modno] = metadata.get(modno, dict())
+            return
+
+        cal_id_to_name = {
+            self.calibration_id(calibration_name): calibration_name
+            for calibration_name in calibrations}
+        calibration_ids = list(cal_id_to_name.keys())
+
+        resp_versions = CalibrationConstantVersion.get_closest_by_time_by_detector_conditions(  # noqa
+            self.client,
+            self.detector_name,
+            calibration_ids,
+            condition.api_encode(self),
+            karabo_da=next(iter(da_to_modno)) if len(da_to_modno) == 1 else '',
+            event_at=self._api_format_time(event_at),
+            snapshot_at=self._api_format_time(snapshot_at))
+
+        if not resp_versions['success']:
+            raise CalCatError(resp_versions)
+
+        for ccv in resp_versions['data']:
+            try:
+                modno = da_to_modno[ccv['physical_detector_unit']['karabo_da']]
+            except KeyError:
+                # Not included in our modules
+                continue
+
+            cc = ccv['calibration_constant']
+
+            # Manually build the result dict with expressive keys.
+            entry = dict(
+                cc_id=cc['id'],
+                cc_name=cc['name'],
+                condition_id=cc['condition_id'],
+                ccv_id=ccv['id'],
+                ccv_name=ccv['name'],
+                path=Path(ccv['path_to_file']) / ccv['file_name'],
+                dataset=ccv['data_set_name'],
+                begin_validity_at=ccv['begin_validity_at'],
+                end_validity_at=ccv['end_validity_at'],
+                raw_data_location=ccv['raw_data_location'],
+                start_idx=ccv['start_idx'],
+                end_idx=ccv['end_idx'])
+
+            metadata.set_pdu_ccv(
+                modno, cal_id_to_name[cc['calibration_id']], entry)
+
+
+class DualConditionCalibrationData(CalibrationData):
+    """Calibration data with dark and illuminated distinction.
+
+    Some detectors of this kind distinguish between two different
+    operating conditions depending on whether photons illuminate the
+    detector or not, correspondingly called the illuminated and dark
+    conditions. Typically the illuminated condition is a superset of the
+    dark condition.
+
+    Not all implementations for semiconductor detectors inherit from
+    this type, but only those that make this distinction such as AGIPD
+    and LPD.
+    """
+
+    dark_calibrations = set()
```

Comment:
Comparison: in `calng`, each detector-specific subclass can define arbitrary methods to generate conditions and the attribute `_constants_need_conditions` explicitly maps from constants (or "calibrations") to condition methods. One can of course translate between that and `dark_calibrations` / `illuminated_calibrations`, so it's in the end a matter of modeling things nicely.
Personally, I find the explicit map easier to read (to figure out what goes into getting an `Offset`, I see that I need to look at `self.dark_condition`). Additionally, it avoids the introduction of a "dual condition" subclass.

Reply:
I'm still somewhat hopeful we can get rid of the split between dark and illuminated condition. It makes the interaction unnecessarily complicated and with tolerances, there's a better and native way to express this distinction anyway (e.g. dark constants are defined with `source_energy = 0 +/ inf`).

Apart from that, given there are de facto only those two conditions (and only for some detectors anyway), I'm not sure it warrants an entire additional layer of abstraction across all detectors (said mapping). With a subclass, I can make the distinction easily disappear later on.